### ISSUE TITLE
Event.to_hash sometimes contains anonymous modules or classes, which cannot be dumped to YAML

### ISSUE THREAD

**Issue by nateberkopec** [getsentry/raven-ruby#511]
It's common to have code like this:

```ruby
SentryJob = Struct.new(:event) do
  def perform
    Raven.send_event(event)
  end
end

Raven.configure do |config|
  config.async = lambda { |event|
    Delayed::Job.enqueue SentryJob.new(event.to_hash)
  }
end
```

`event.to_hash` can contain an anonymous module, which cannot be dumped to YAML formats for use in a job processor.

**Comment by dcramer** [comment]
It might make sense for us to change how async works so we encode the object before passing it in.

**Comment by nateberkopec** [comment]
JSON backends seem to work fine (any storage adapter that calls `to_json` instead of `to_yaml`).

**Comment by nateberkopec** [comment]
We could just hand the encoded version to the job processor, though that will probably break everyones existing jobs.

**Comment by nateberkopec** [comment]
Confirmed that #506 is a duplicate of this. The error message there is slightly different: 

> ActiveJob::SerializationError: Unsupported argument type: Class

**Comment by nateberkopec** [comment]
I think it's Rails that's doing this (something in the Rack env, perhaps), so I'm going to have to try to get a repro case.

**Comment by nateberkopec** [comment]
Also it would be good to try to fallback to sync'd event sending if something in `config.async` raises an exception, as suggested.

**Comment by nuclearpidgeon** [comment]
I am experiencing a similar issue trying to async report with ActiveJob, using DelayedJob as the backend.

I get an `Unsupported argument type: Time` error that comes out of [activejob-4.2.6/lib/active_job/arguments.rb:69](https://github.com/rails/rails/blob/v4.2.6/activejob/lib/active_job/arguments.rb#L69)

Our job code at the moment is pretty much the same as the Sentry example:

```
class ReportSentryEventJob < ActiveJob::Base
  queue_as :default

  def perform(event)
    Raven.send_event(event)
  end
end
```
```
# config/production.rb
  Raven.configure do |config|
    # Process sending exceptions using ActiveJob
    config.async = lambda do |event|
      ReportSentryEventJob.perform_later(event.to_hash)
    end
  end
```

I logged the `event.to_hash` hash to take a look at it and spotted a few Time/DateTime values nested under the 'extra' key. Note the lack of string quotes around the time values:

```
:extra => {
  :delayed_job => {
    :id => 766,
    :priority => 0,
    :attempts => 2,
    :run_at => Tue, 20 Sep 2016 14:07:29 AEST +10:00,
    :locked_at => Tue, 20 Sep 2016 14:07:34 AEST +10:00,
    :locked_by => "host:mystery-hostname.net.au pid:60742",
    :queue => "default",
    :created_at => Tue, 20 Sep 2016 14:05:26 AEST +10:00
  },
  :last_error => "Unsupported argument type: Time\n/Users/stewartwebb/.rvm/gems/ruby-2.2.3@webapp/gems/activejo",
  :handler => "--- !ruby/object:ActiveJob::QueueAdapters::DelayedJobAdapter::JobWrapper\njob_data:\n  job_class: OurA",
  :active_job => {
    "job_class" => "OurAppJob",
    "job_id" => "ca2ba326-769d-4868-a826-9f1cb7ed63fc",
    "queue_name" => "default",
    "arguments" => [
      {
        "_aj_globalid" => "gid://webapp/OurAppDatabaseModel/91"
      }
    ],
    "locale" => :en
  }
}
```

(OurAppJob/OurAppDatabaseModel are just redacted names)

Looking at the DelayedJob integration file [sentry-raven-1.2.3/lib/raven/integrations/delayed_job.rb:14](https://github.com/getsentry/raven-ruby/blob/v1.2.3/lib/raven/integrations/delayed_job.rb#L14), I can see that a few Time/DateTime objects just get passed straight through from the job (run_at, locked_at etc).

I added a `.to_s` call to all the '_at' variables, which fixed the `Unsupported argument type: Time` error but raised another: `Can't serialize a Hash with reserved key \"_aj_globalid\"`
(see [activejob-4.2.6/lib/active_job/arguments.rb:126](https://github.com/rails/rails/blob/v4.2.6/activejob/lib/active_job/arguments.rb#L126) )

Adding a `.inspect` call to the end of the line that adds data into the `:active_job`-keyed hash ([sentry-raven-1.2.3/lib/raven/integrations/delayed_job.rb:33](https://github.com/getsentry/raven-ruby/blob/v1.2.3/lib/raven/integrations/delayed_job.rb#L33) gets rid of that error, but also gets rid of the nice formatting/pretty-printing that Sentry does in the issue view.

I have put together a PR with the above fixes for convenience, but there is potentially a better way to handle this - figured it was best to just leave it in Sentry's hands

**Comment by nateberkopec** [comment]
Argh, that's pretty bad that this happens with our DJ adapter and ActiveJob. Thanks for the detailed writeup.

Before passing the event to `async`, we could walk the event data and try to [coerce everything into the ActiveJob-compatible types](https://github.com/rails/rails/blob/master/activejob/lib/active_job/arguments.rb#L38).

The only alternative I see is passing the JSON-encoded version (so, just a big String), so that all the job processor does is send and wait on an HTTP call. That's a breaking change for anyone that modifies the event somehow in `async` or prior to calling `send_event`.

Both approaches increase the amount of work done *prior* to calling `async`.

**Comment by nateberkopec** [comment]
Closed in #548, #547.
